### DISCUSSION TITLE
IMessageInspector replacement/missing, something different I need to do?

### DISCUSSION THREAD

**Discussion by ThisBytes5** [CoreWCF/CoreWCF#911] [Q&A] [answered]
in .net 4.8 we had a behavior that used the below IdentityMessageInspector to read a message header that our clients add to all of our WCF requests. This controls how we know who the user is and what data they have access too so it is a requirement for us to continue to move to CoreWCF (currently using SoapCore for our first 4.8 --> .net core move).

I've spent last couple days searching for a solution and not finding anything (could be bad search terms). 
Is there a replacement for this type of function that I'm missing or is IMessageInspector the only way to do it and it just hasn't been implemented yet?

I'm fine with changing how we process the header since we are migrating, but I can't really change the fact that the data we need is passed in the header.

`  public class IdentityMessageInspector : IMessageInspector
  {
    public object AfterReceiveRequest(ref Message message, ServiceDescription serviceDescription)
    {
      int headerIndex = message.Headers.FindHeader("IdentityHeader", "http://landstar.com/IdentityHeader");
`

**Comment by gplwhite** [comment]
I think you want to use IDispatchMessageInspector. This interface is the extensibility point for interacting with messages on the dispatch (ie: server) side. 

IClientMessageInspector is the equivalent extensibility point for interacting with messages on the client side.


In your situation, you would have an implementation of IClientMessageInspector that adds the identity header to messages sent from the client. And you'd have an implementation of IDispatchMessageInspector that looks for the identity header on the server side. (I'm doing something similar and it works fine CoreWCF.

Your IdentityMessageInspector class could implement both interfaces if you want it to perform both sides of the inspection. 
However in practice I've found it more useful to have separate implementations to avoid the client taking a dependency on CoreWCF (IClientMessageInspector is available from the Wcf Client Library (ie: System.ServiceModel), but IDispatchMessageInspector is only available from CoreWCF).

**Reply by ThisBytes5** [reply]
> I think you want to use IDispatchMessageInspector. This interface is the extensibility point for interacting with messages on the dispatch (ie: server) side.

Thanks for the reply, will give this a shot!

**Reply by ThisBytes5** [reply]
> IDispatchMessageInspector

Any chance you have an example of how to use this? I spent most of the day yesterday attempting to get it to work, but no luck.

What I've done:
IdentityDispatchMessageInspector : IDispatchMessageInspector <-- processes the message (see below), but receive exception when attempting to get the Header XML.

IdentityEndpointBehavior : IEndpointBehavior <-- Registers the IDispatchMessageInspector 

public class IdentityServiceBehavior : IServiceBehavior <-- This goes through each endpoint and addes the IdentityEndpointBehavior to each endpoint.


headerText below always equals: CoreWCF.Channels.HeaderInfoCache+HeaderInfo 
which obviously isn't XML. I've attached the Identity header in case it's helpful.

`
    public object AfterReceiveRequest(ref Message message, IClientChannel channel, InstanceContext instanceContext)
    {
#if DEBUG
      Console.WriteLine("HERE");
#endif

      int headerIndex = message.Headers.FindHeader("IdentityHeader", "http://landstar.com/IdentityHeader");
      if (headerIndex != -1)
      {
        string headerText = message.Headers[headerIndex].ToString();
        headerText = headerText.Replace("xmlns=\"http://landstar.com/IdentityHeader\"", string.Empty);

        //Shouldn't happen, if header isn't present then headerIndex would have been -1
        if (XMLTools.DeserializeXml(headerText, typeof(IdentityHeader)) is not IdentityHeader header) **<-- Exception here due to invalid XML**
          throw new InvalidIdentityHeaderException("Unable to deserialize IdentityHeader");
  }
`



[IdentityHeader.txt](https://github.com/CoreWCF/CoreWCF/files/10067513/IdentityHeader.txt)

**Comment by gplwhite** [comment, marked as answer]
Hi,

Arguably my implementation is a bit simpler as I use simple string values for each header, rather than a fully serialized object like you are doing. But at first glance I don't see any major issue with deriving from MessageHeader and overriding the serialization/deserialization like you're doing.
However I am a bit confused why you're trying to Deserialize the header manually in the message inspector above - as you've encapsulated the deserialization logic inside your custom message header. I think you should just be able to retrieve the message header as your custom message header type and the framework will take care of the deserialization by calling your deserialization overload.

I think you should just be able to do something like this:

```
int headerIndex = message.Headers.FindHeader("IdentityHeader", "http://landstar.com/IdentityHeader");
if (headerIndex != -1)
{
    IdentityHeader identity = headers.GetHeader<IdentityHeader>(headerIndex);
}
```


**EDIT** Sorry - didn't pay enough attention to your code. I didn't realise only the serialization is handled by the MessageHeader overload. I'll take another look, but I'm guessing there will be somewhere to plug in the deserialization logic too.

**EDIT2** This is a bit of a stab but you could try changing your IdentityHeader class to be a simple poco - don't derive from MessageHeader. And instead decorate it with DataContractAttribute and the members with DataMemberAttribute.
It appears that the message header infrastructure is using a DataContractSerializer internally, so I suspect it will serialize/deserialize without you having to explicitly control the serialization/deserialization.   
When you want to add an instance of the header on the client side, you would do this:
`headers.Add(MessageHeader.CreateHeader("IdentityHeader", "http://landstar.com/IdentityHeader", identityHeader))`
and reading the header on the dispatch side would be as in my example above using GetHeader<>()

**Reply by ThisBytes5** [reply]
Got everything converted to a poco, all attributes added, and after some fighting with the DataContract Serializer (had to add order to the datamembers) everything on the server is working. 

Now going to do the Client side and see if I can get that working with the new stuff or just leave it alone.

Thank you very much!
